File: Serene/Serene.Script/Common/Helper/ExcelExportHelper.cs

namespace Serene.Common
{
    using Serenity;
    using System;
    using System.Runtime.CompilerServices;

    [Imported]
    public static class ExcelExportHelper
    {
        public static ToolButton CreateToolButton(ExcelExportOptions options)
        {
            return null;
        }
    }

    [Imported, Serializable]
    public class ExcelExportOptions
    {
        public Serenity.IDataGrid Grid { get; set; }
        public string Service { get; set; }
        public Func<bool> OnViewSubmit { get; set; }
        public string Title { get; set; }
    }

}

File: Serene/Serene.Web/Modules/Common/Helpers/ExcelExportHelper.ts
namespace Serene.Common {

    export interface ExcelExportOptions {
        grid: Serenity.DataGrid<any, any>;
        service: string;
        onViewSubmit: () => boolean;
        title?: string;
        hint?: string;
        separator?: boolean;
    }

    export namespace ExcelExportHelper {

        export function createToolButton(options: ExcelExportOptions): Serenity.ToolButton {

            return {
                hint: Q.coalesce(options.title, 'Excel'),
                title: Q.coalesce(options.hint, ''),
                cssClass: 'export-xlsx-button',
                onClick: function () {
                    if (!options.onViewSubmit()) {
                        return;
                    }

                    let grid = options.grid;

                    var request = Q.deepClone(grid.getView().params) as Serenity.ListRequest;
                    request.Take = 0;
                    request.Skip = 0;
                    var sortBy = grid.getView().sortBy;
                    if (sortBy) {
                        request.Sort = sortBy;
                    }

                    request.IncludeColumns = [];
                    let columns = grid.getGrid().getColumns();
                    for (let column of columns) {
                        request.IncludeColumns.push(column.id || column.field);
                    }
                    Q.postToService({ service: options.service, request: request, target: '_blank' });
                },
                separator: options.separator
            };
        }
    }
}